### PR TITLE
[Stable10] [WIP] Federated sharing new spec

### PR DESCRIPTION
**Pull request by VicDeo**
## Description
Backport of 
https://github.com/owncloud/core/pull/32086
https://github.com/owncloud/core/pull/32113
https://github.com/owncloud/core/pull/32303

- [x] `/ocm-provider` endpoint
- [x] discovery of `/ocm-provider` endpoint
- [x] fallback to the old endpoint when `/ocm-provider` is not accessible
- [x] `/index.php/apps/federatedfilesharing/shares` to process new incoming shares
- [x] `/index.php/apps/federatedfilesharing/notifications` to process new incoming notifications
- [ ] sending/recieving the of the following notifications
  - [x] `SHARE_ACCEPTED`
  - [x] `SHARE_DECLINED`
  - [x] `SHARE_UNSHARED`
  - [x] `REQUEST_RESHARE` 
  - [ ] `RESHARE_UNDO` (not tested yet)
  - [x] `RESHARE_CHANGE_PERMISSION`
## Related Issue
- Fixes <issue_link>

## Motivation and Context
Implements OCM API 1.0-proposal1

###  Environment: 10.0.8 (**stable**)  and this branch (**ocm**)
- [x] sharing from **stable** + declining from **ocm**
- [x] sharing from **stable** + accepting from  **ocm** + unsharing from  **ocm**
- [x] sharing from **stable** + accepting from  **ocm** + unsharing from   **stable**

- [x] sharing from **ocm** + declining from **stable**
- [x] sharing from **ocm** + accepting from  **stable** + unsharing from  **stable**
- [x] sharing from **ocm** + accepting from  **stable** + unsharing from   **ocm**

###  Environment:  two instances @  this branch - **ocm1** and **ocm2**
- [x] sharing from **ocm1** + declining from **ocm2**
- [x] sharing from **ocm1** + accepting from  **ocm2** + unsharing from  **ocm2**
- [x] sharing from **ocm1** + accepting from  **ocm2** + unsharing from   **ocm1**

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:
- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
- [x] Add more unit tests
- [ ] Think over how new permissions will be stored (change DAV to expose OCM permissions if needed)
- [ ] expose to the user Display names of sender/owner/etc
- [ ] expose to the user Optional share description
- [ ] exposed to the user `message` field in new notifications 

## Nice to haves (once the spec leaves it's current draft status):
- [ ] Refactor the old code to the new spec completely leaving only RequestHandlerController in place. 
It would just convert old data to match the new requirements
E.g. `OCA\FederatedFileSharing\BackgroundJob\RetryJob` still uses data in old format.

### CODE REVIEW

File: apps/federatedfilesharing/lib/DiscoveryManager.php [hunk 1]
```diff
@@ -90,39 +90,105 @@ private function discover($remote) {
 		if (\defined('PHPUNIT_RUN') && !$this->underTest) {
 			return $discoveredServices;
 		}
-		// Read the data from the response body
+
+		$decodedService = $this->makeRequest($remote . '/ocs-provider/');
```
Comment (tomneedham) [inline apps/federatedfilesharing/lib/DiscoveryManager.php]: should be `ocm-provider` ?

Comment (VicDeo) [reply apps/federatedfilesharing/lib/DiscoveryManager.php]: no, this is a fallback endpoint, for OCM see line #141

**Comment by tomneedham** [comment]
Supplied to some partners for testing - waiting for feedback :)

**Comment by PVince81** [comment]
conflict @VicDeo

**Comment by tomneedham** [comment]
We received some feedback, with two instances using this PR, with 2.5.0 and 2.5.1 client, that revoking a folder share causes some issues.

**Comment by PVince81** [comment]
General note: it is likely that bugs already existed in the way the storages were handled, independently from the OCS / OCM federation protocol. We'll need to keep this in mind when evaluating whether said issues are related to the move to the new OCM version or older federation bugs.

**Comment by VicDeo** [comment]
Conflicts are resolved

**Comment by PVince81** [comment]
as discussed in chat, this is ready for merge

**Comment by PVince81** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by ownclouders** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Pull request merged**